### PR TITLE
Submitting Lessons 03, 04, 05, 06

### PR DESCRIPTION
**Pull request by cthaehtruthteller**
This pull request is for:

- Lesson03: strformat.py, list_lab.py, mailroom.py

- Lesson04: mailroom_part2.py

- Lesson05: except_exercise.py, mailroom_part3.py

- Lesson06: test_mailroom_part4.py, mailroom_part4.py

### CODE REVIEW

**Comment by natasha-aleksandrova** [comment]
@cthaehtruthteller just a heads up, lesson03/mailroom will get reviewed first and if there are issues that need addressed then lesson04, 05 and 06 of the mailroom will be on hold pending updates. 

The mailroom assignments builds up each lesson so it's imperative to have it reviewed one step at a time.

File: students/thecthaeh/Lesson04/mailroom_part2.py [hunk 1]
```diff
@@ -60,32 +58,47 @@ def handle_selection2():
...
     
     #print the sorted report
     for donor in sorted_report:
         print("{:30}  ${:>19.2f}  {:>15}  ${:>19.2f}".format(*donor[:]))
     
     prompt()
 
+#function for selection 3
+def option3():
+    letter_dict = {}
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson04/mailroom_part2.py]: why create this dictionary? you can do the same thing in just looping over the main data structure:
```
for name, donations in donor_info_dict.items():
    f.write("...".format(name, sum(donations))
```

**Comment by cthaehtruthteller** [comment]
I went through the comments and resubmitted mailroom_part2.

**Comment by natasha-aleksandrova** [comment]
@cthaehtruthteller great! will take another look at again tonight or tomorrow morning.

File: students/thecthaeh/Lesson04/mailroom_part2.py [hunk 2]
```diff
@@ -60,32 +55,36 @@ def handle_selection2():
...
+    actions = prompt()
+    while True:
+        if actions == '4' or actions == 'quit':
+            break
+        menu = {
+            1: single_thank_you, 
+            2: create_report,
+            3: thank_all
+        }
+        actions = menu.get(int(actions))()
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson04/mailroom_part2.py]: You are still calling "main" within other functions. Your "feature" functions should be unaware of the main flow.
See reference to this in the tutorial: https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html


couple other comments:
* your switch dict mapping should be outside of while loop
* promt (input()) function should be inside while loop
* quit options should be mapped to a function which calls sys.exit()


so your code should look like this:
```
menu = {"1": single_thank_you, "2": create_report, "3": thank_all, "4": quit_program, "quit": "quit_program}
while True:
    response = input("...")
    menu[response]()

```
that's it!

File: students/thecthaeh/Lesson04/trigrams.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    for x in range(len(a_seq) - 2):
+        pair = tuple(a_seq[x:x + 2]) # will be the keys in trigrams. try frozen sets instead of tuples
+        follower = a_seq[x + 2] # will be the values in trigrams
+
+        #try to update this to use defaultdict(list) instead
+        if pair in trigrams:
+            trigrams[pair].append(follower)
+        else:
+            trigrams[pair] = [follower]
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson04/trigrams.py]: no need to make changes here; just wanted to note to check out `collections.defaultdict` for using dictionaries with lists as value.

File: students/thecthaeh/Lesson04/trigrams.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def generator():
+    #TESTING: update first word pair in new_text to be determined randomly & test with longer string
+    new_text = ["I", "wish"]
+
+    while tuple(new_text[-2:]) in trigrams:
+        new_text.append(random.choice(trigrams[tuple(new_text[-2:])]))
+    
+    return new_text
+
+def main():
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson04/trigrams.py]: is this assignment complete? this is a syntax error here and you should be reading and writing to a file as part of this work.

File: students/thecthaeh/Lesson05/except_exercise.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+
+# Figure out what the exception is, catch it and while still
+# in that catch block, try again with the second item in the list
+first_try = ['spam', 'cheese', 'mr death']
+
+try:
+    joke = fun(first_try[0])
+except NameError:
+    joke = fun(first_try[2])
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson05/except_exercise.py]: second item in the list is "cheese" which would be `first_try[1]`

File: students/thecthaeh/Lesson05/except_exercise.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+langs = ['java', 'c', 'python']
+
+try:
+    more_joke = more_fun(langs[0])
+except IndexError:
+    more_joke = more_fun(langs[1])
+else:
+    more_joke = more_fun(langs[2])
+finally:
+    last_fun()
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson05/except_exercise.py]: looks good, besides minor indexing issue (comment above)

File: students/thecthaeh/Lesson04/mailroom_part2.py (hunk 2 above)
Comment (cthaehtruthteller) [reply students/thecthaeh/Lesson04/mailroom_part2.py]: Ah. So none of my functions should need to call prompt()?

Comment (natasha-aleksandrova) [reply students/thecthaeh/Lesson04/mailroom_part2.py]: Exactly!

your prompt() or input() directly would just be once inside of a while True loop

**Comment by cthaehtruthteller** [comment]
I resubmitted mailroom_part2, trigrams, and except_exercise.

**Comment by cthaehtruthteller** [comment]
I also just updated mailroom_part3 and part4 and am resubmitting those.

File: students/thecthaeh/Lesson05/mailroom_part3.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    menu = {
+            '1': single_thank_you, 
+            '2': create_report,
+            '3': thank_all,
+            '4': quit_program,
+            'quit': quit_program
+        }
+    while True:
+        prompt = input("Choose one of the following options: \n1. Send a Thank You to a single donor \n2. Create a Report \n3. Send letters to all donors \n4. Quit \n>>> ")
+        menu[prompt]()
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson05/mailroom_part3.py]: this part needs exception handling - what happens when user enters "5" ? 

(hint: catch KeyError here)

File: students/thecthaeh/Lesson05/mailroom_part3.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+def create_report():
+    """Create a report listing donors, their total donation amount, the number of donations, and their average donation."""
+    header = ['Donor Name', 'Total Given', 'Num Gifts', 'Average Gifts']
+    
+    top_row = "{:30} |{:>20} |{:>15} |{:>20}".format(*header[:])
+    print(top_row)
+    print('-' * len(top_row))
+    
+    report = [[donor, sum(donor_info_dict[donor]), len(donor_info_dict[donor]), sum(donor_info_dict[donor])/len(donor_info_dict[donor])] for donor in donor_info_dict]
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson05/mailroom_part3.py]: you can make this line much shorter and more readable if you use items() (which is preferred anyway)
```
report = [[name, sum(donations), len(donations), sum(donations)/len(donations) for name, donations in donor_info_dict.items()]
```

File: students/thecthaeh/Lesson06/test_mailroom_part4.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    assert x == donors.keys()
+    
+    #test for a new donor: if the donor is new, the donor is added to the dict
+    add_donor("Benny Hana")
+    assert donors["Benny Hana"] == []
+
+#listing donors
+def test_list_donors():
+    list = []
+    for donor in donors: list.append(donor)
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson06/test_mailroom_part4.py]: this should be on two separate lines

File: students/thecthaeh/Lesson06/test_mailroom_part4.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    noah_row = ['Noah Bunch', 10.00, 2, 5.00]
+    pony_row = ['Pony Skittle', 21.00, 3, 7.00]
+    assert report(test_donor) == [top_row, [pony_row, noah_row]]
+
+#Tests for Thank You text files
+def test_text_file():
+    #text_file(donors)
+    #test file creation (one file per donor)
+    for donor in donors:
+        os.path.exists(f"./{donor}.txt")
```
Comment (natasha-aleksandrova) [inline students/thecthaeh/Lesson06/test_mailroom_part4.py]: each test function should be calling your code - there is also no assert here?

what is the purpose of the above block?

**Pull request merged**